tools/codegen/api/types.py

```diff
@@ -1,145 +1,116 @@
 from tools.codegen.model import *
 from dataclasses import dataclass
-from typing import Optional, Union, Sequence, Tuple, TypeVar, List
+from typing import Optional, Union, Sequence, TypeVar, List
+from enum import Enum
 
 _T = TypeVar('_T')
 
-# ------------------------------------------------------------------- #
+# An ArgName is just the str name of the argument in schema;
+# but in some special circumstances, we may add a little extra
+# context.  The Enum SpecialArgName covers all of these cases;
+# grep for their construction sites to see when they can occr.
+
+SpecialArgName = Enum('SpecialArgName', (
+    'possibly_redundant_memory_format',
+))
+ArgName = Union[str, SpecialArgName]
+
+# A CType is short for C++ semantic type.  A CType represents a C++ type, plus
+# semantic information about what it represents.  For example, consider the
+# argument "bool pin_memory"; its normal C++ type is "bool", but its C++
+# semantic type also keeps track that this represents a "pin_memory"; you can't
+# just use a random other boolean in a context where you need a "pin_memory"!
+#
+# CTypes encode C++ type structure as needed for translation.  Right now we
+# track references and optional, but don't, for example, track ArrayRef.  If
+# you need trnsnlations that know about these types, beef up this data
+# structure.
 
-#                           cpp types
+@dataclass(frozen=True)
+class BaseCType:
```

Comment:
I find the term "semantic type" confusing, I liked "argument" better. Or actually, a type plus a name is a "parameter" rather.

Reply:
The bigest trouble with the name Argument is that CTypes also accurately classify C++ expressions, which are definitely not arguments. (Although, I'm a little confused where you attached this comment: Cpp/DispatcherArgument got renamed to Binding, but I'm not opposed to renaming Binding back to Argument, except that conflicts with model.Argument which already took this name. CArgument would be in line with CType naming convention.